### PR TITLE
Fix WinHttpHandler response handling for manual cookie parsing

### LINKED ISSUE #4960: HttpClient not sending cookies which were part of earlier redirect response in Core Clr
I am using CoreClr version of HttpClient on my Windows 10 OS.

**Scenario:** Successful login will redirect the user to a page where the logging in user info is displayed

**Issue:** Since the necessary cookie is not present, the system thinks that the user is not logged in

**Trace of calls for the above scenario(Note: all these messages were logged at the server side)**

---
Request: *request for login page*
GET /account/login ||Connection: Keep-Alive||Accept-Encoding: gzip, deflate||Cookie: U58hr9PQBAQ=CfDJ8IcTvLBrmYtHm19jNmsLC62rbm_jC2LBGRmurk0fd2R9YeGdeOLdRpC-rEVSSK24hsMjOToMUnjgo_cZwFLrDge_a0sHVy5eJc9qYkjYvThiFcG7MA9MB-HYLsoxlQIzT6v8tmAUbTOZ-39AakEWLTI; .AspNet.Session=7e8cee49-8d6b-6ab5-357b-d29e964af60d||Host: localhost:5000

Response: *login page sent to client*
200 GET /account/login ||Date: Sat, 12 Dec 2015 14:27:39 GMT||Transfer-Encoding: chunked||Content-Type: text/html; charset=utf-8||Server: Kestrel||

Request: *client request to login with details of username and password*
POST /Account/Login ||Connection: Keep-Alive||Content-Length: 69||Content-Type: application/x-www-form-urlencoded||Accept-Encoding: gzip, deflate||Cookie: U58hr9PQBAQ=CfDJ8IcTvLBrmYtHm19jNmsLC62rbm_jC2LBGRmurk0fd2R9YeGdeOLdRpC-rEVSSK24hsMjOToMUnjgo_cZwFLrDge_a0sHVy5eJc9qYkjYvThiFcG7MA9MB-HYLsoxlQIzT6v8tmAUbTOZ-39AakEWLTI; .AspNet.Session=7e8cee49-8d6b-6ab5-357b-d29e964af60d||Host: localhost:5000||

Response: *server successful login response with redirect to home page /*
302 POST /Account/Login ||Cache-Control: no-cache||Date: Sat, 12 Dec 2015 14:27:39 GMT||Pragma: no-cache||Expires: -1||Location: /||Server: Kestrel||Set-Cookie: .AspNet.Microsoft.AspNet.Identity.Application=CfDJ8IcTvLBrmYtHm19jNmsLC60z9IhWh_D65HPs01cDjGLzpfs5B5SeoX4k5tem5BqxDavdDWtijmOSUKuWw_lkI3PEeWJjUHn_mk8ArChjrFgySrEuCSkQMZU4vGVL3R3F2_aNbpixFkI3Oh2_PN1VF9Peb1XUJJTZDCDLlUFP7QSreFC4qaiJWaWTpGdp-_nStQSr2OzP5qAvqbGn2ME4fICTfm7BqWas_h_WaZPWjOwRT68V9ewibMxR2odTbcQf3Bqie1IEx2kjwiYLdXrzaT2sc-TmzKnNSTELgWmCVBdV27dMM2YGlwUOIL_YTgyAqCNLSO_YjfG8_gTODvhjTCJ9ff8hc_xTx0PYiZMdah5oNZPmAyAC2FJh_hWCRQD1jzgDVVGp2le9TaEw6XPdvsa_2u1HbjH_vRBIdtd49i9ZJv9Z_rO5X3PDgTYkAahAlGx7ym_vDOTZA-VC3KvM-JMOyIfcjYjovU9QkhsXBg8UI5COArVxLPXHSyXADoltftMxOg8oKBdl-55I8bKe-0ep6haIAz6ozlSFR41fEQN_j9kyNCaSjZYGu6-A6Qu8YP6tGEb3EPGhicE15NtrkDj7YhnJruOGKip2ZMvv-V0xg1UfzmNgOvvpMZHElbFsL2jlPaP0nzPn4uffArK6hlxo_ExYOdp596omQjPCF4h-; path=/; httponly||

**NOTE:** This call is missing the `.AspNet.Microsoft.AspNet.Identity.Application` cookie

Request: *client request to home page*
GET / ||Connection: Keep-Alive||Accept-Encoding: gzip, deflate||Cookie: U58hr9PQBAQ=CfDJ8IcTvLBrmYtHm19jNmsLC62rbm_jC2LBGRmurk0fd2R9YeGdeOLdRpC-rEVSSK24hsMjOToMUnjgo_cZwFLrDge_a0sHVy5eJc9qYkjYvThiFcG7MA9MB-HYLsoxlQIzT6v8tmAUbTOZ-39AakEWLTI; .AspNet.Session=7e8cee49-8d6b-6ab5-357b-d29e964af60d||Host: localhost:5000||

Response:
200 GET / ||Date: Sat, 12 Dec 2015 14:27:39 GMT||Transfer-Encoding: chunked||Content-Type: text/html; charset=utf-8||Server: Kestrel||

---

**NOTE** that this scenario works fine with regular desktop HttpClient.

@Tratcher

### PR DESCRIPTION
**Pull request by davidsh**
WinHttpHandler was losing 'Set-Cookie' response headers during auto redirection handling when the handler was in 'manual cookie' mode. I.e. `CookieUsePolicy.UseSpecifiedCookieContainer`. This is the only mode that HttpClientHandler uses.

The 'Set-Cookie' headers were currently only parsed out when a final HTTP response happened. This meant that 'Set-Cookie' headers present in interim 3xx redirection responses as well as interim 401, 407 auth responses would not be stored into the CookieContainer. Hence, those same cookies could not be sent back to the server when needed.

The fix is to make sure we parse out the 'Set-Cookie' headers whenever we get an HTTP response back from WinHTTP. As part of this fix, I re-factored out the cookie handling code into a separate class. I also added specific WinHttpHandler tests for the two different cookie modes as well as a HttpClientHandler test to validate cross-platform (i.e. CurlHandler).

Fixes #4960

### CODE REVIEW

**Comment by davidsh** [comment]
@CIPop @stephentoub PTAL

File: src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpCookieHelper.cs [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+            Debug.Assert(cookies != null);
+
+            string cookieValues = cookies.GetCookieHeader(uri);
+            if (!string.IsNullOrEmpty(cookieValues))
+            {
+                cookieHeader = string.Format(
+                    CultureInfo.InvariantCulture,
+                    "{0} {1}",
+                    CookieHeaderNameWithColon,
+                    cookieValues);
```
Comment (justinvp) [inline src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpCookieHelper.cs]: `cookieHeader = HttpKnownHeaderNames.Cookie + ": " + cookieValues;` should be more efficient than `string.Format` and avoids having to redefine the `Cookie` constant.

File: src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpResponseParser.cs [hunk 2]
```diff
@@ -130,7 +109,39 @@ public static uint GetResponseHeaderNumberInfo(SafeWinHttpHandle requestHandle,
...
+            uint index = 0;
+            
+            return GetResponseHeaderStringHelper(requestHandle, infoLevel, ref index);
+        }
+
+        // Returns all headers or an empty list if header isn't found.
+        public static List<string> GetResponseHeaders(SafeWinHttpHandle requestHandle, uint infoLevel)
+        {
+            uint index = 0;
+            var headerList = new List<string>();
```
Comment (justinvp) [inline src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpResponseParser.cs]: This is always going to allocate a new `List<string>` (and underlying `string[]` inside the list), even if there aren't any response headers.

File: src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpResponseParser.cs [hunk 3]
```diff
@@ -130,7 +109,39 @@ public static uint GetResponseHeaderNumberInfo(SafeWinHttpHandle requestHandle,
...
+            string header;
+            
+            while (true)
+            {
+                header = GetResponseHeaderStringHelper(requestHandle, infoLevel, ref index);
```
Comment (justinvp) [inline src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpResponseParser.cs]: Each time `GetResponseHeaderStringHelper` is called a new `StringBuilder` is allocated (along with the `char[]` inside the builder). It'd be nice to be able to reuse the buffer across calls so there aren't so many allocations on each loop iteration, which is what #3199 was aiming to address for previous uses of `GetResponseHeaderStringInfo`.

File: src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpCookieHelper.cs [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+using System.Globalization;
+using System.Runtime.InteropServices;
+
+using SafeWinHttpHandle = Interop.WinHttp.SafeWinHttpHandle;
+
+namespace System.Net.Http
+{
+    internal static class WinHttpCookieHelper
+    {
+        private const string CookieHeaderNameWithColon = "Cookie:";
```
Comment (justinvp) [inline src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpCookieHelper.cs]: `private const string CookieHeaderNameWithColon = HttpKnownHeaderNames.Cookie + ":";`?

File: src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpCookieHelper.cs (hunk 1 above)
Comment (davidsh) [inline src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpCookieHelper.cs]: Fixed. I had forgotten about the `Common/src/System/Net/HttpKnownHeaderNames.cs` class.

File: src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpCookieHelper.cs (hunk 4 above)
Comment (davidsh) [inline src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpCookieHelper.cs]: Fixed.

File: src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpResponseParser.cs (hunk 3 above)
Comment (davidsh) [inline src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpResponseParser.cs]: Good observations. I realize that #3199 is going to optimize some of these memory allocations.  And that pending PR can address this as well. Thx for your patience with that pending PR!

For now, this cookie fix is high priority to address the ASP.NET DNX scenario which is broken right now.

File: src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpResponseParser.cs (hunk 2 above)
Comment (davidsh) [inline src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpResponseParser.cs]: Do you have specific suggestions for an optimization here?  For now, we can let #3199 try and micro-optimize those allocs.

Comment (justinvp) [inline src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpResponseParser.cs]: You could only allocate the list when there is a header to add, and either return null or an empty list for no headers. I'm not a huge fan of returning null, but it might be easiest here and there are other methods that already return null for no headers, e.g. `GetResponseHeaderStringInfo` when the header isn't found (callers would have to check for null before foreaching over the list). Returning a static readonly empty list would also work as the list is only used internally, so we know nobody is going to add to it (and asserts could be added to ensure it stays empty).

Something like:

```c#
public static List<string> GetResponseHeaders(SafeWinHttpHandle requestHandle, uint infoLevel)
{
    uint index = 0;
    List<string> headerList = null;
    string header;
    
    while (true)
    {
        header = GetResponseHeaderStringHelper(requestHandle, infoLevel, ref index);

        if (header != null)
        {
            if (headerList == null)
            {
                headerList = new List<string>();
            }

            headerList.Add(header);
        }
        else
        {
            break;
        }
    }
    
    return headerList;
}
```

Comment (davidsh) [inline src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpResponseParser.cs]: These are good ideas. For now, I'll keep this PR as-is but welcome further optimizations later.

**Comment by davidsh** [comment]
Test Innerloop OpenSUSE13.2 Debug Build and Test

File: src/System.Net.Http.WinHttpHandler/src/System.Net.Http.WinHttpHandler.msbuild [hunk 5]
```diff
@@ -15,6 +15,7 @@
     <CompileItem Include="$(CommonPath)\Interop\Windows\winhttp\Interop.winhttp.cs" />
     <CompileItem Include="$(CommonPath)\Microsoft\Win32\SafeHandles\SafeHandleZeroOrMinusOneIsInvalid.cs" />
     <CompileItem Include="$(CommonPath)\System\StringExtensions.cs" />
+    <CompileItem Include="$(CommonPath)\System\Net\HttpKnownHeaderNames.cs" />
```
Comment (CIPop) [inline src/System.Net.Http.WinHttpHandler/src/System.Net.Http.WinHttpHandler.msbuild]: This adds a lot of unnecessary strings. We should at least have a TODO item saying that we need to split it into partial classes to ensure that only the needed strings are added to the binary.

File: src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpCookieHelper.cs [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+        public static void AddResponseCookiesToContainer(WinHttpRequestState state)
+        {
+            HttpRequestMessage request = state.RequestMessage;
+            SafeWinHttpHandle requestHandle = state.RequestHandle;
+            CookieContainer cookieContainer = state.Handler.CookieContainer;
+            
+            Debug.Assert(state.Handler.CookieUsePolicy == CookieUsePolicy.UseSpecifiedCookieContainer);
+            Debug.Assert(cookieContainer != null);
+
+            // Get 'Set-Cookie' headers from response
```
Comment (CIPop) [inline src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpCookieHelper.cs]: Nit: missing "."

File: src/System.Net.Http.WinHttpHandler/src/System.Net.Http.WinHttpHandler.msbuild (hunk 5 above)
Comment (davidsh) [inline src/System.Net.Http.WinHttpHandler/src/System.Net.Http.WinHttpHandler.msbuild]: FWIW, the System.Net.Http.dll binary already drags in this class `HttpKnownHeaderNames.cs`.  Now this change will drag in another copy for the System.Net.Http.WinHttpHandler.dll as well.

But won't the compiler assist in this in some way by figuring out what is really needed and trim out the rest?

Comment (justinvp) [inline src/System.Net.Http.WinHttpHandler/src/System.Net.Http.WinHttpHandler.msbuild]: FYI, in #3199 I extended `HttpKnownHeaderNames` to allow looking up and reusing the header string instances to avoid having to always allocate new strings for known response headers. Not sure how easy it will be to split apart the strings with that optimization -- maybe splitting it up between request vs. response headers at some point.

File: src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpCookieHelper.cs (hunk 6 above)
Comment (davidsh) [inline src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpCookieHelper.cs]: Fixed.

File: src/System.Net.Http.WinHttpHandler/src/System.Net.Http.WinHttpHandler.msbuild (hunk 5 above)
Comment (davidsh) [inline src/System.Net.Http.WinHttpHandler/src/System.Net.Http.WinHttpHandler.msbuild]: Although we're using a little more static storage in the library, doing too much micro-optimization on these allocations might end up making this code much harder to maintain and extend in the future. So, we should balance that against all of these optimizations (which we don't really understand the true quantitative benefits of).

Comment (CIPop) [inline src/System.Net.Http.WinHttpHandler/src/System.Net.Http.WinHttpHandler.msbuild]: >  So, we should balance that against all of these optimizations (which we don't really understand the true quantitative benefits of).

What is the DLL size before and after this change?

My proposal was to simply create HttpKnownHeaderNames.Cookie.cs and HttpKnownHeaderNames.cs as partial and import within each project only what is needed. The split will be made whenever it's necessary to reduce the DLL size.

File: src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpCookieHelper.cs [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+            foreach (string cookieHeader in cookieHeaders)
+            {
+                WinHttpTraceHelper.Trace(cookieHeader);
+                try
+                {
+                    cookieContainer.SetCookies(request.RequestUri, cookieHeader);
+                }
+                catch (CookieException)
+                {
+                    // We ignore malformed cookies in the response.
```
Comment (CIPop) [inline src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpCookieHelper.cs]: Shouldn't we somehow identify malformed content and report it back to the application? Is this the Desktop behavior?

File: src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpCookieHelper.cs [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+                }
```
Comment (CIPop) [inline src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpCookieHelper.cs]: We should at least trace that there was a malformed cookie in the response.

File: src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpCookieHelper.cs (hunk 7 above)
Comment (davidsh) [inline src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpCookieHelper.cs]: This is the Desktop behavior and also the behavior for WinHTTP and WinINET.  They tend to be resilent in the face of garbage on the wire and simply drop malformed cookies.

File: src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpCookieHelper.cs [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Runtime.InteropServices;
+
+using SafeWinHttpHandle = Interop.WinHttp.SafeWinHttpHandle;
+
+namespace System.Net.Http
+{
+    internal static class WinHttpCookieHelper
```
Comment (CIPop) [inline src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpCookieHelper.cs]: This doesn't really feel like a static class: most functions keep getting the state object. Why not construct a new object instead.

Also, the name isn't conforming to the single responsibility principle: in my mind a "helper" is good for anything so it'll end up getting all sorts of functions doing various things with cookies over the time. The actual purpose seems to be adapting from WinHttp to .Net CookieContainer: maybe renaming to CookieContainerAdapter would make the intent clear?

File: src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpCookieHelper.cs [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+                    cookieHeader,
+                    (uint)cookieHeader.Length,
+                    Interop.WinHttp.WINHTTP_ADDREQ_FLAG_ADD))
+                {
+                    WinHttpException.ThrowExceptionUsingLastError();
+                }
+            }
+        }
+        
+        public static string GetCookieHeader(Uri uri, CookieContainer cookies)
```
Comment (CIPop) [inline src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpCookieHelper.cs]: Should be private.

File: src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs [hunk 11]
```diff
@@ -620,7 +604,7 @@ internal static string GetCookieHeader(Uri uri, CookieContainer cookies)
             // Manually add cookies.
             if (cookies != null)
             {
-                string cookieHeader = GetCookieHeader(requestMessage.RequestUri, cookies);
+                string cookieHeader = WinHttpCookieHelper.GetCookieHeader(requestMessage.RequestUri, cookies);
```
Comment (CIPop) [inline src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs]: This code could be moved within the Cookie adapter class above to encapsulate cookie processing within a single class.

File: src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpCookieHelper.cs (hunk 8 above)
Comment (davidsh) [inline src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpCookieHelper.cs]: Fixed.

File: src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpResponseParser.cs [hunk 12]
```diff
@@ -86,31 +86,10 @@ internal static class WinHttpResponseParser
...
-                            // We ignore malformed cookies in the response.
-                        }
-                    }
-                }
-            }
-
             return response;
         }
 
+        // Returns the first header or throws if header isn't found.
```
Comment (CIPop) [inline src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpResponseParser.cs]: Nit: there was a discussion asking CoreFX developers to use (or keep existing) `///` documentation-style XML comments.

File: src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpCookieHelper.cs (hunk 10 above)
Comment (davidsh) [inline src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpCookieHelper.cs]: Fixed.

File: src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpResponseParser.cs (hunk 12 above)
Comment (davidsh) [inline src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpResponseParser.cs]: There is nothing currently published in the official guidelines yet for this that I'm aware of.  Can you point me to "discussion" or document?

File: src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.cs [hunk 13]
```diff
@@ -385,11 +385,31 @@ public void MaxRequestContentBufferSize_Set_ThrowsPlatformNotSupportedException(
...
                 }
             }
         }
 
+        [ActiveIssue(4960, PlatformID.AnyUnix)]
+        [Theory]
+        [InlineData("cookieName1", "cookieValue1")]
+        public async Task GetAsync_RedirectResponseHasCookie_CookieSentToFinalUri(string cookieName, string cookieValue)
+        {
+            Uri uri = HttpTestServers.RedirectUriForDestinationUri(false, HttpTestServers.RemoteEchoServer, 1);
```
Comment (CIPop) [inline src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.cs]: Please add more details to what this does. What's "false" and 1 in this context? (Maybe you could use the parameter names within the call to make things a bit clearer.)

File: src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpResponseParser.cs [hunk 14]
```diff
@@ -167,7 +178,7 @@ private static string GetResponseHeaderStringInfo(SafeWinHttpHandle requestHandl
                 Interop.WinHttp.WINHTTP_HEADER_NAME_BY_INDEX,
```
Comment (CIPop) [inline src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpResponseParser.cs]: Nit: Above: `bytesNeeded / 2` should be `bytesNeeded / sizeof(char)`.

File: src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpCookieHelper.cs [hunk 15]
```diff
@@ -0,0 +1,95 @@
...
+            List<string> cookieHeaders =
+                WinHttpResponseParser.GetResponseHeaders(requestHandle, Interop.WinHttp.WINHTTP_QUERY_SET_COOKIE);
+            WinHttpTraceHelper.Trace("WINHTTP_QUERY_SET_COOKIE");
+            
+            foreach (string cookieHeader in cookieHeaders)
+            {
+                WinHttpTraceHelper.Trace(cookieHeader);
+                try
+                {
+                    cookieContainer.SetCookies(request.RequestUri, cookieHeader);
```
Comment (CIPop) [inline src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpCookieHelper.cs]: Will the request.RequestUri be the correct filter URI if automatic redirection is used?

File: src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpResponseParser.cs (hunk 12 above)
Comment (CIPop) [inline src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpResponseParser.cs]: I remember seeing an email. There's an issue regarding public documentation here: https://github.com/dotnet/corefx/issues/230
It would be nice to understand what to use for internal function descriptions.

File: src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpCookieHelper.cs (hunk 15 above)
Comment (davidsh) [inline src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpCookieHelper.cs]: request.RequestUri is updated during auto redirection in the WinHttpStatusCallback.  So, it will have been updated already before this code above runs in the case where this method is called after a terminal response.  

In the case where it is called during the actual redirection callback, then it uses the current URI (not the new one yet) because it needs to put cookies into the container based on the the current request being processed.

So, in both cases, the correct URI is used.

File: src/System.Net.Http.WinHttpHandler/src/System.Net.Http.WinHttpHandler.msbuild (hunk 5 above)
Comment (davidsh) [inline src/System.Net.Http.WinHttpHandler/src/System.Net.Http.WinHttpHandler.msbuild]: I think for now, I'm going to undo the inclusion of `HttpKnownHeaderNames.cs` from this PR and simply use "Cookie" string directly.

I'll let the next PR such as #3199 look into further optimizations.

File: src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.cs (hunk 13 above)
Comment (davidsh) [inline src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.cs]: "false" is the `secure` parameter
"1" is the number of `hops` (redirection)

Changed some of the method calls to use named arguments.

File: src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpResponseParser.cs (hunk 14 above)
Comment (davidsh) [inline src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpResponseParser.cs]: Fixed.

File: src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpCookieHelper.cs (hunk 9 above)
Comment (davidsh) [inline src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpCookieHelper.cs]: I'll change the name of the class to `WinHttpCookieContainerAdapter`.  Longer term, I plan to refactor this code and use a nested class, separate file pattern, similar to CurlHandler.  I.e., make these "helper" classes subclasses of WinHttpHandler but split into separate source files, i.e. WinHttpHandler.CookieContainerAdapter.cs, etc.  This would also be more consist with the project coding guidelines.

However, I don't want to do a "new" of this class.  The `state` object is created for each request.  But the logic of the "adapter" is fixed.  So, it actually makes sense to pass in a `state` object to the methods.

I don't want to block on this for this PR.  But if there are other ideas, I can consider for next time.

File: src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs (hunk 11 above)
Comment (davidsh) [inline src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs]: The calculation of the 'Cookie' header value is already in the adapter class.  Moving the logic of appending the request header itself seems like a stretch.  Then the adapter has to know about appending headers in general which seems out of scope.

**Comment by davidsh** [comment]
Test Innerloop Windows_NT Release Build and Test

File: src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpCookieHelper.cs (hunk 10 above)
Comment (CIPop) [inline src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpCookieHelper.cs]: Right: I left a comment there as well.

File: src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs (hunk 11 above)
Comment (CIPop) [inline src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs]: It's already doing a lot of header changes: it's calling `WinHttpAddRequestHeaders` after all.

File: src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpCookieHelper.cs (hunk 15 above)
Comment (CIPop) [inline src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpCookieHelper.cs]: Clarified offline: WinHttp will call the callback even if automatic redirects are used.

**Comment by CIPop** [comment]
LGTM

The object model needs improvement.
We should assert that the WinHttp callback flags are such that the redirect callback is _always_ called. Otherwise we run the risk of exposing cookies to unwanted websites during a redirect.

**Comment by davidsh** [comment]
Added comments and issue #5036 to track the WinHttp callback flags improvements.

**Comment by davidsh** [comment]
Test  Innerloop Ubuntu Release Build and Test
